### ISSUE TITLE
NormalizerInterface type mismatch between supportsNormalization and normalize

### ISSUE THREAD

**Issue by luka8088** [symfony/symfony#27457]
**Symfony version(s) affected**: 4.1.0

**Description**  
NormalizerInterface has two methods: `supportsNormalization` which accepts a value of any type and `normalize` which accepts only objects. In case `supportsNormalization` returns `true` for a non-object value the same value will be passed to `normalize` thus violating the interface.

**How to reproduce**  
```php
$normalizer = new class implements \Symfony\Component\Serializer\Normalizer\NormalizerInterface {

    /**
     * {@inheritdoc}
     */
    public function normalize($object, $format = null, array $context = array())
    {
        // @see https://github.com/symfony/serializer/blob/079905bcd7/Normalizer/NormalizerInterface.php#L28
        // As per docs `$data` is `object`.
        assert(is_object($object), 'Value of type `' . gettype($object) . '` received where object expected.');
        return serialize($object);
    }

    /**
     * {@inheritdoc}
     */
    public function supportsNormalization($data, $format = null)
    {
        // @see https://github.com/symfony/serializer/blob/079905bcd7/Normalizer/NormalizerInterface.php#L44
        // As per docs `$data` is `mixed`.
        // Returning `true` here for a non-object value
        // will cause a non-object value to be passed to `::normalize`.
        return true;
    }

};

$serializer = new \Symfony\Component\Serializer\Serializer(
  [$normalizer],
  [new \Symfony\Component\Serializer\Encoder\JsonEncoder()]
);

// @see https://github.com/symfony/serializer/blob/079905bcd7/SerializerInterface.php#L24
$serializer->serialize(['foo'], 'json');
```

**Possible Solution**  
Maybe make the Serializer assume that all non-object values are not supported for normalization?

**Additional context**
Found using https://gitlab.com/phlint/phlint while using the Serializer in another project.

```
  ✖ Argument Compatibility: $data in vendor/symfony/serializer/Serializer.php:133
    Argument #1 passed in the expression `$normalizer->normalize($data, $format, $context)` is of type `array`.
    A value of type `array` is not implicitly convertible to type `object`.
      Trace #1:
        #1: Method `public function normalize(string[] $data, 'json' $format, array $context)`
          specialized for the expression `$this->normalize($data, $format, $context)`
          in `vendor/symfony/serializer/Serializer.php:106`.
        #2: Method `public final function serialize(string[] $data, 'json' $format, array $context)`
          specialized for the expression `$serializer->serialize(['foo'], 'json')`
          in `sample.php:39`.
    Rule documentation: https://gitlab.com/phlint/phlint/blob/master/documentation/rule/argumentCompatibility.md
```

**Comment by xabbuh** [comment]
IMO this is a mistake in your normalizer's `supportsNormalization()` method is your return `true` for a value that you then do not handle properly in `normalize()`.

**Comment by luka8088** [comment]
Hey @xabbuh! Even if I put `return is_string($data);` instead of `return true;` and change `normalize` to handle strings I would still receive a `string` where the interface says I should receive an `object` - and that is the problem this ticket is about.

**Comment by nicolas-grekas** [comment]
We could trigger a deprecation when supportsNormalization returns true while a non-object is passed to it. That would allow us to add the `object` type hint on the `normalize()` methods.
(but honestly, that's a very minor issue IMHO, I'd be very fine leaving things as is, a docblock is just a hint.)

**Comment by andrewnclark** [comment]
I've made a start on this issue and took your advice to trigger a deprecation when a non-object is passed into supportsNormalization and we still get true. In order to add the object type hint the composer requirements must change.

Object type hints are only available in PHP 7.2+, doing so causes two tests to fail in JsonSerializableNormalizerTest.php - do you want to work around these and them them green, or just provide a deprecation with the intent to upgrade PHP at a later date?

**Comment by vudaltsov** [comment]
@nicolas-grekas , we cannot add `object` type hint to `NormalizerInterface::normalize()` because it is supposed to process any given value recursively. See its implementation in [`Serializer::normalize()`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Serializer/Serializer.php#L141).

I agree with `mixed` typehint idea.
